runtime/tests/org.eclipse.core.tests.runtime/src/org/eclipse/core/tests/runtime/jobs/JobGroupTest.java

```diff
@@ -1293,6 +1310,76 @@ public IStatus run(IProgressMonitor monitor) {
 		assertTrue("2.0", numShouldCancelCalled[0] < NUM_JOBS_LIMIT + 10);
 	}
 
+	/**
+	 * Tested scenario: - 2 jobs are started in parallel and the first one returns a
+	 * CANCEL_STATUS while the second one is still running. This lets
+	 * <code>JobGroup.shouldCancel</code> return <code>true</code> which triggers
+	 * the cancelation of any running job.
+	 *
+	 * Expected result: The second job is explicitly canceled
+	 *
+	 */
+	public void testShouldCancel_5() {
+
+		// the job group allows for 2 threads so both jobs will be started in
+		// parallel
+		final JobGroup jobGroup = new JobGroup("JobGroup", 2, 2) {
+			@Override
+			protected boolean shouldCancel(IStatus lastCompletedJobResult, int numberOfFailedJobs,
+					int numberOfCanceledJobs) {
+				// cancel the group as soon as one job in it returns the CANCEL_STATUS
+				return numberOfCanceledJobs > 0;
+			}
+		};
+
+		final TestBarrier2 returnsCancelStatusBarrier = new TestBarrier2();
+		Job returnsCancelStatusJob = Job.create("The one that cancels", __ -> {
+			returnsCancelStatusBarrier.setStatus(TestBarrier2.STATUS_WAIT_FOR_RUN);
+			returnsCancelStatusBarrier.waitForStatus(TestBarrier2.STATUS_RUNNING);
+
+			return Status.CANCEL_STATUS;
+		});
+
+		final TestBarrier2 canceledJobBarrier = new TestBarrier2();
+		AtomicBoolean jobWasCanceled = new AtomicBoolean();
+		Job canceledJob = new Job("The one that is explicitly canceled") {
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				canceledJobBarrier.setStatus(TestBarrier2.STATUS_WAIT_FOR_RUN);
+				canceledJobBarrier.waitForStatus(TestBarrier2.STATUS_RUNNING);
+
+				// give it time: this Job needs to be running at the time the other job finishes
+				canceledJobBarrier.waitForStatus(TestBarrier2.STATUS_DONE);
+				return Status.OK_STATUS;
+			}
+
+			@Override
+			protected void canceling() {
+				jobWasCanceled.set(true);
+				canceledJobBarrier.setStatus(TestBarrier2.STATUS_DONE);
+			}
+		};
+
+		returnsCancelStatusJob.setJobGroup(jobGroup);
+		returnsCancelStatusJob.schedule();
+
+		canceledJob.setJobGroup(jobGroup);
+		canceledJob.schedule();
+
+		// Wait until both jobs are ready to run and let them start running
+		returnsCancelStatusBarrier.waitForStatus(TestBarrier2.STATUS_WAIT_FOR_RUN);
+		canceledJobBarrier.waitForStatus(TestBarrier2.STATUS_WAIT_FOR_RUN);
+
+		returnsCancelStatusBarrier.setStatus(TestBarrier2.STATUS_RUNNING);
+		canceledJobBarrier.setStatus(TestBarrier2.STATUS_RUNNING);
+
+		// give it time: the job needs to be canceled
+		canceledJobBarrier.waitForStatus(TestBarrier2.STATUS_DONE);
+
+		assertTrue("The second job should have been explicitly canceled", //
+				jobWasCanceled.get());
```

Comment:
This is basically dead code. This assertion can only fail  if the `waitForStatus` call fails and in that case the test already ends after that call. So `jobWasCanceled` is obsolete, as it is already checked through the barrier.

Reply:
I removed the assertion and added a comment that makes it clear, that reaching the end of the test method means that the test passed.
FWIW, I know it was dead code, I just find it more expressive to have assertions and variables with meaningful names in a test, I think it better expresses the purpose of the test.

Reply:
I see the point. But then there should be a documentation that states why the code is there despite being unnecessary (from a functional perspective). Otherwise, someone maintaining the test will wonder what this (dead) code is supposed to do.